Add custom domain support

zbee opened the pull request:
(Maintainer's note: moved and merged via #108)

I added support for custom domains in `ApiConfig`, with the region appended as a query parameter, and provided an example in the ReadMe.

All behavior is the same if you don't pass a custom domain, but otherwise it forms the called URLs with your custom domain + the endpoint + the region as a query parameter (with consideration to the few endpoints that have parameters already).
This fits [the recommended proxy](https://hextechdocs.dev/using-cloudflare-workers-to-make-api-calls/) format.

Finally, I also updated the ReadMe's example code to use the the correct `ApiConfig` and `GetChampionMasteries`.

(apologies for the ReadMe entire change, it matched line-endings on commit)

zbee committed:
Support custom domains

zbee committed:
Fix example code in readme

zbee committed:
Add example for a custom API domain

zbee committed:
Remove unused config variable "apiRegionAsSubdomain"

zbee commented:
@MingweiSamuel Just wanted to try to bring this up your queue again.
I have now tested it with the Account and 2 Match endpoints - one of which uses query parameters.

MingweiSamuel commented on src/Camille.RiotGames/gen/configspec.json:
```diff
@@ -3,6 +3,11 @@
     "type": "string",
     "desc": "Riot Games API key."
   },
+  "apiURL": {
```
`apiUrl`

MingweiSamuel commented on src/Camille.RiotGames/src/Util/RegionalRequester.cs:
```diff
@@ -37,13 +35,40 @@ public class RegionalRequester
         /// </summary>
         private readonly HttpClient _client = new HttpClient();
 
+        /// <summary>
+        ///  If true, the region will be appended to the route instead of used as a subdomain.
+        /// </summary>
+        private bool _regionAsSubdomain;
```
Make this a configuration setting (in configspec) instead of inferred

zbee replied:
Moved to `Camille.RiotGames.apiCallRegionConfig`, which uses a `Camille.RiotGames.RegionConfig` enum value.

MingweiSamuel commented on src/Camille.RiotGames/src/Util/RegionalRequester.cs:
```diff
@@ -67,6 +92,18 @@ public RegionalRequester(IRiotGamesApiConfig config, string route)
             var num429s = 0;
             var num5xxs = 0;
+
+            // If the region is not being used as a subdomain
+            if (!_regionAsSubdomain)
+            {
+                // Append the region as a query parameter, with consideration for other parameters
+                var uri = request.RequestUri.ToString()
+                          + (!request.RequestUri.ToString().Contains("?") ? "?" : "&")
+                          + $"region={_region}";
+                // Replace the request with a new one that has the region correctly appended
+                request = new HttpRequestMessage(request.Method, uri);
+            }
```
Should be able to eliminate this extra logic if `regionAsSubdomain` is in the configspec

zbee replied:
Unfortunately: no.
The bulk of the logic around line 101 in your snippet (now [line 105](https://github.com/MingweiSamuel/Camille/blob/f3906c7d8895415792ce48194bf7990a6e5eaa2c/src/Camille.RiotGames/src/Util/RegionalRequester.cs#L105)) is instead based off of whether the endpoint uses query parameters based on the method's optional parameters - but since those parameters are baked into the URL instead of specified as `request.Properties` (I think?) I don't see how this could get much cleaner.

---

Edit: Outdated, see [this further request](https://github.com/MingweiSamuel/Camille/pull/103#discussion_r1520413210) or [its resolving commit](https://github.com/MingweiSamuel/Camille/pull/103/commits/ed06e6a86360bee69b60212cc8c6755987fd3870).

MingweiSamuel commented on README.md:
````diff
@@ -1,215 +1,224 @@
+        // ...
+    }.Build()
+);
+// OR with a proxy server
+var riotApi = RiotGamesApi.NewInstance(
+    new RiotGamesApiConfig.Builder("")
+    {
+        // ...
+        ApiURL = "proxy.myexample.app",
+    }.Build()
+);
+// Which will send requests like https://proxy.myexample.app/lol/summoner/v4/summoners/by-name/NAME?region=NA1
+```
````
Doesn't really need to be top of README, should be pretty obvious what it does from the docs

MingweiSamuel reviewed (changes requested):
Ok to merge this I'd just want to move the configuration of subdomain vs query param into the config as well instead of guessing based on the url. (Bonus points if you want to make it also handle other ways like a path segment or header but not really needed)

zbee commented:
Thank you so much for reviewing this!
I will look to get these changes implemented tonight.

I'll try to find other proxy scripts and see if I can support how they need the region passed, good suggestion.

MingweiSamuel commented:
Yeah in Riven the url is configured with a format param `{}`:

https://github.com/MingweiSamuel/Riven/blob/8d810ccca86b29c86dd4eb4891bc4060faf8c4d3/riven/src/config.rs#L26-L29

Which works for subdomain and path prefixes, but not for query params

```rust
    /// `"https://{}.api.riotgames.com"`
    ///
    /// Default base URL, including `{}` placeholder for region platform.
    pub const DEFAULT_BASE_URL: &'static str = "https://{}.api.riotgames.com";
```

zbee committed:
Removed proxy documentation

zbee committed:
Updated test away from stale data

zbee committed:
Rename ApiURL variable to fit code style

zbee committed:
Add support for region as a header parameter, and more specific URL building

zbee commented:
I have updated this, closing all but one of your change requests above.

I implemented the below suggestion, and an `enum` to choose how the region should be applied to the API call, to make it easier to support the region as a path parameter in the future, but I only directly added support for the region to be a header (and the existing query parameter and subdomain, of course).

> Yeah in Riven the url is configured with a format param `{}`:

Thank you again for your review, and I welcome any further change requests of course!

AoshiW commented on src/Camille.RiotGames/gen/RiotGamesApiConfig.cs.dt:
```diff
@@ -9,6 +9,17 @@ using System;
 
 namespace Camille.RiotGames
 {
+    public enum RegionConfig
+    {
+        /// <summary>The Region should be in the URL, as a subdomain</summary>
+        inUrlAsSubdomain,
+        /// <summary>The Region should be in the URL, as a query parameter</summary>
+        inUrlAsRegionQueryParameter,
+        /// <summary>The Region should be a header instead of in the URL</summary>
+        /// <seealso cref="RiotGamesApiConfig.RegionHeaderKey"/>
+        inHeader,
+    }
```
name of enum values should start with a capital letter
(and maybe it could be moved from `gen` to `src` - this enum is not generated)

zbee replied:
Thank you, and both items are now implemented.
I had missed that consideration in my thoughts of keeping it alongside the configuration variables themselves.

zbee committed:
Moved RegionConfig to Camille.RiotGames.Enums, and changed its values casing

zbee commented:
@MingweiSamuel Just wanted to bump this up again, I implemented your and AoshiW's change requests barring one.

MingweiSamuel commented on src/Camille.RiotGames/gen/configspec.json:
```diff
@@ -3,6 +3,21 @@
   },
+  "apiUrl": {
+    "type": "string",
+    "desc": "URL to call for the Riot Games API. Include {0} for the region, if you want it in the URL - Not necessary for RegionConfig.inUrlAsRegionQueryParameter.",
```
"Not necessary if apiCallRegionConfig is RegionConfig.InUrlAsRegionQueryParameter"

MingweiSamuel commented on src/Camille.RiotGames/gen/configspec.json:
```diff
@@ -3,6 +3,21 @@
+    "desc": "How the API call should include the region.",
+    "val": "RegionConfig.InUrlAsSubdomain"
+  },
+  "regionHeaderKey": {
+    "type": "string",
+    "desc": "The header key to use for the region. Only used if RegionConfig is set to inHeader.",
+    "val": "\"Riot-Region\""
+  },
```
Could you make this work with `InUrlAsRegionQueryParameter` too and just call it `ReigonKey`? 

"The query or header name to use for the region. Only used if RegionConfig is set to InHeader or InQuery."

MingweiSamuel commented on src/Camille.RiotGames/src/Util/RegionalRequester.cs:
```diff
@@ -37,13 +36,44 @@ public class RegionalRequester
+                _client.DefaultRequestHeaders.Add(config.RegionHeaderKey, route);
+            }
+
+            // The API key is only needed for riot's API, otherwise it is assumed to be a keyed proxy
+            if (_config.ApiUrl.Contains("riotgames.com"))
+            {
+                _client.DefaultRequestHeaders.Add(RiotKeyHeader, config.ApiKey);
+            }
```
Would prefer to not do any branching based on searching the ApiUrl - could either make this another config field like `includeRiotApiHeader` --

Or if the user is using a proxy they could just set the config.ApiKey to a meaningless value and the proxy can ignore it (maybe a bit sloppy but should work fine)

MingweiSamuel commented on src/Camille.RiotGames/src/Util/RegionalRequester.cs:
```diff
@@ -67,6 +97,18 @@ public RegionalRequester(IRiotGamesApiConfig config, string route)
+            // If the region is not being used as a subdomain
+            if (_config.ApiCallRegionConfig == RegionConfig.InUrlAsRegionQueryParameter)
+            {
+                // Append the region as a query parameter, with consideration for other parameters
+                var uri = request.RequestUri
+                          + (!request.RequestUri.ToString().Contains("?") ? "?" : "&")
+                          + $"region={_region}";
```
Would be better to do something like https://stackoverflow.com/a/14517976/2398020 to ensure everything is properly escaped (wow microsoft, this is really clunky)

zbee replied:
Yes, it is clunky ha - made only clunkier by System.Web's "absence" in a few of the target frameworks.
I explain further in a comment below.

MingweiSamuel commented on src/Camille.RiotGames/src/Enums/RegionConfig.cs:
```diff
@@ -0,0 +1,18 @@
+﻿// MobaHinted Copyright (C) 2024 Ethan Henderson <ethan@zbee.codes>
+// Licensed under GPLv3 - Refer to the LICENSE file for the complete text
+
```
Presumably this was autogenerated - definitely need to remove this and also double check that it's ok to release your PR under the existing Apache2 + MIT licenses?

MingweiSamuel reviewed (commented):
Thank you for your patience, really appreciate this PR, just a few small changes to do and otherwise looks good!

MingweiSamuel commented:
Apologies for the slow response time, just missed the notifications

zbee committed:
Remove generated license callout

zbee committed:
Correct variable in apiUrl description

zbee committed:
Clarify apiUrl description

zbee committed:
Move RegionHeaderKey to instead be RegionKey, and use it for the query parameter

zbee committed:
Use the new RegionKey variable

zbee committed:
Do not branch off of the URL contents

zbee committed:
Use RegionKey to build the URL without string checking

zbee commented on src/Camille.RiotGames/Camille.RiotGames.csproj:
```diff
@@ -17,6 +17,18 @@
     <Service Include="{508349b6-6b84-4df5-91f0-309beebad82d}" />
   </ItemGroup>
 
+  <ItemGroup>
+    <Reference Include="System.Web" Condition="'$(TargetFramework)' == 'net45'">
+      <HintPath>$(NuGetPackageRoot)\microsoft.netframework.referenceassemblies.net45\1.0.3\build\.NETFramework\v4.5\System.Web.dll</HintPath>
+    </Reference>
+    <Reference Include="System.Web" Condition="'$(TargetFramework)' == 'net461'">
+      <HintPath>$(NuGetPackageRoot)\microsoft.netframework.referenceassemblies.net461\1.0.3\build\.NETFramework\v4.6.1\System.Web.dll</HintPath>
+    </Reference>
+    <Reference Include="System.Web" Condition="'$(TargetFramework)' == 'netcoreapp3.1'">
+      <HintPath>$(NuGetPackageRoot)\microsoft.netcore.app.ref\3.1.0\ref\netcoreapp3.1\System.Web.dll</HintPath>
+    </Reference>
+  </ItemGroup>
```
I did what I could to keep the targets the same, but ultimately had to remove `netstandard1.1` and add references for `net45`, `net461`, and `netcoreapp3.1`.

It seems [this could be done more cleanly](https://stackoverflow.com/a/5710941/1843510) than what I have done in `Camille.RiotGames.csproj`, however none of those options exist for me in Rider or VS. Maybe they do for you, or it makes more sense to you?

MingweiSamuel replied:
Hah good, always need an excuse to knock off the old versions

zbee commented:
No worries at all about any slowness, I very much appreciate the time you've taken on this.
I have completed your requested changes now, with one concern noted above, which doesn't seem to me to be something I am able to address further myself.

MingweiSamuel commented on src/Camille.RiotGames/src/Util/RegionalRequester.cs:
```diff
@@ -67,6 +95,19 @@ public RegionalRequester(IRiotGamesApiConfig config, string route)
+                query[_config.RegionKey] = _region;
+                uri.Query = query.ToString();
+                // Replace the request with a new one that has the region correctly appended
+                request = new HttpRequestMessage(request.Method, uri.ToString());
```
This would actually break any body params, (or header params if they exist, maybe with RSO, IDK)

MingweiSamuel reviewed (commented):
Made some changes myself, in PR #108 and will merge in a sec

MingweiSamuel commented:
Commits squashed and moved to #108
